Prevent duplicate extensions

cwillisf opened the pull request:
### Resolves

Resolves #694 

### Proposed Changes

This change prevents two extensions / extension instances from being registered under the same extension ID (from `getInfo`). For internal extensions this is relatively easy: we just unregister the service when we find out it's a duplicate. For a worker-based extension, though, we don't find out about the duplicate until the extension worker has started up and the extension object has been instantiated in that worker.

Since we can't be sure that this all happens with predictable timing -- especially if an extension doesn't register before the extension framework starts up -- I chose to handle the situation with a watchdog. Any duplicate extension ID registration coming from a worker is ignored, and if no extension has been successfully registered by a worker for a specific amount of time (currently 5 seconds) then the worker is terminated. This also guards against some probably-would-never-happen-but-maybe edge cases, such as two extensions being registered from one worker where one of the extensions duplicates the ID of a third extension, provided by another worker.

If the GUI knows an extension ID, it can call `isExtensionLoaded` to short-circuit all of the above at the UI layer.

### Reason for Changes

Without these changes it's possible to load multiple copies of the same extension, leading to duplicate blocks and mass hysteria.

rschamp reviewed (changes requested):
Off thread we discussed splitting this into a separate PR for the "watchdog" functionality, and simplifying the dedupe functionality to prevent registration altogether by checking against which extension URLs have been loaded.

Commit:
Add an init watchdog for extension workers

If an extension worker does not complete initialization in a reasonable
amount of time (currently 15 seconds) then the worker will be
terminated. A complete initialization includes registering an extension.

Commit:
Refuse to register second extension with same ID

If an extension attempts to register with the same ID as another
extension which has already registered, the new registration is refused.
If the extension is in a worker and no other extension is successfully
registered by that worker, the watchdog system will terminate the
"empty" worker.

Commit:
Simplify extension duplication prevention station

Since we will only have builtin extensions for a while, the
more-complicated, watchdog-based extension duplication prevention has
been replaced with one that treats the extension URL as equivalent to
the extension ID. This is true only for built-in extensions, so once we
start supporting third-party extensions we will likely need to bring
back the watchdog (commit e4381b4693dbed9746992c9a5b3be73b078a2171) and
the more-complex duplication prevention (commit
670e51d33580e8a2e852b3b038bb3afc282f81b9).

The code here should be treated as a last resort: ideally the GUI should
check if a particular extension is loaded (maybe by calling the
Extension Manager's new `isExtensionLoaded` method) and avoid calling
`loadExtensionURL` if it is.

cwillisf commented:
I've updated this branch and the PR description with a simpler VM-side duplication prevention mechanism.

@rschamp, since we probably don't need the watchdog system for now do you think it would still make sense to make a separate PR for that? I'm happy to have it hang out in the Git history like this -- I added a comment with the relevant commit hash so if/when we want to bring it back, we can.

Commit:
Add built-in extension "URL" to _loadedExtensions

Dupe prevention works better when you actually record the necessary
info...

ericrosenbaum reviewed (approved):
It works!
